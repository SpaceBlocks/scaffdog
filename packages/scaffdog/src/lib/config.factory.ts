import merge from 'deepmerge';
import { isPlainObject } from 'is-plain-object';
import type { ResolvedConfig } from '@scaffdog/types';
import type { PartialDeep } from 'type-fest';

export const createResolvedConfig = (
  props: PartialDeep<ResolvedConfig> = {},
): ResolvedConfig =>
  merge(
    {
      files: [],
      variables: new Map(),
      helpers: new Map(),
    },
    props as any,
    {
      isMergeableObject: isPlainObject,
    },
  );
