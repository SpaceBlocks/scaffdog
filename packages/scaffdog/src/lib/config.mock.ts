// NOTE: This is auto-generated file.
import { vi } from 'vitest';
import type { ConfigLibrary } from './config';
export const createConfigLibraryMock = (
  mocks: { [P in keyof ConfigLibrary]?: any } = {},
): ConfigLibrary => ({
  load: vi.fn(),
  ...mocks,
});
