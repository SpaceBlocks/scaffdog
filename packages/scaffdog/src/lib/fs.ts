import { readFileSync, statSync } from 'fs';
import { mkdir, readFile, writeFile } from 'fs/promises';
import type { GlobbyOptions } from 'globby';
import globby from 'globby';

export type FsMkdirOptions = {
  recursive?: boolean;
};

export type FsGlobOptions = GlobbyOptions;

export type FsLibrary = {
  mkdir: (filepath: string, options: FsMkdirOptions) => Promise<void>;
  readFile: (filepath: string) => Promise<string>;
  readFileSync: (filepath: string) => string;
  glob: (
    patterns: string | readonly string[],
    options: FsGlobOptions,
  ) => Promise<string[]>;
  writeFile: (filepath: string, data: string) => Promise<void>;
  fileExists: (filepath: string) => boolean;
  directoryExists: (filepath: string) => boolean;
};

export const createFsLibrary = (): FsLibrary => {
  const fstat = (filepath: string) => {
    try {
      return statSync(filepath);
    } catch (e) {
      return null;
    }
  };

  return {
    mkdir: async (filepath, options) => {
      await mkdir(filepath, options);
    },

    readFile: async (filepath) => {
      return await readFile(filepath, 'utf8');
    },

    readFileSync: (filepath) => {
      return readFileSync(filepath, 'utf8');
    },

    glob: async (patterns, options) => {
      return await globby(patterns, options);
    },

    writeFile: async (filepath, data) => {
      return await writeFile(filepath, data, 'utf8');
    },

    fileExists: (filepath) => {
      return fstat(filepath)?.isFile() ?? false;
    },

    directoryExists: (filepath) => {
      return fstat(filepath)?.isDirectory() ?? false;
    },
  };
};
