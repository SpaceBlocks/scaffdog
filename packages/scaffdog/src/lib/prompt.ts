/// <reference types="../types/inquirer-autocomplete-prompt" />
import Fuse from 'fuse.js';
import inquirer from 'inquirer';
import inquirerAutocompletePrompt from 'inquirer-autocomplete-prompt';

inquirer.registerPrompt('autocomplete', inquirerAutocompletePrompt);

export type PromptQuestion = inquirer.DistinctQuestion;

export type ConfirmArgs = Omit<PromptQuestion, 'type' | 'name' | 'default'>;

export type AutocompleteArgs = Omit<PromptQuestion, 'type' | 'name'>;

export type PromptLibrary = {
  prompt: <T>(question: PromptQuestion) => Promise<T>;

  confirm: (
    message: string,
    initial: boolean,
    args: ConfirmArgs,
  ) => Promise<boolean>;

  autocomplete: (
    message: string,
    list: string[],
    args: AutocompleteArgs,
  ) => Promise<string>;
};

export const createPromptLibrary = (): PromptLibrary => {
  const prompt: PromptLibrary['prompt'] = async (question) => {
    const { value } = await inquirer.prompt([
      {
        ...question,
        name: 'value',
      },
    ]);

    if (question.when === false && question.default != null) {
      return question.default;
    }

    return value;
  };

  return {
    prompt,

    confirm: async (message, initial, args) => {
      return await prompt({
        ...args,
        type: 'confirm',
        default: initial,
        message,
      });
    },

    autocomplete: async (message, list, args) => {
      const fuse = new Fuse(list, {});

      return await prompt({
        ...args,
        type: 'autocomplete',
        message,
        pageSize: 20,
        source: (_: any, input: string) => {
          return !input
            ? list
            : fuse.search(input || '').map(({ item }) => item);
        },
      } as any);
    },
  };
};
