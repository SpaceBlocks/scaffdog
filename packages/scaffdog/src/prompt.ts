/// <reference types="./types/inquirer-autocomplete-prompt" />
import Fuse from 'fuse.js';
import inquirer from 'inquirer';
import inquirerAutocompletePrompt from 'inquirer-autocomplete-prompt';

inquirer.registerPrompt('autocomplete', inquirerAutocompletePrompt);

export const prompt = async <T>(
  question: inquirer.DistinctQuestion,
): Promise<T> => {
  const { value } = await inquirer.prompt<{ value: T }>([
    {
      ...question,
      name: 'value',
    },
  ]);

  if (question.when === false && question.default != null) {
    return question.default;
  }

  return value;
};

export const confirm = async (
  message: string,
  initial: boolean,
  args: Omit<inquirer.DistinctQuestion, 'type' | 'name' | 'default'> = {},
): Promise<boolean> => {
  return await prompt({
    ...args,
    type: 'confirm',
    default: initial,
    message,
  });
};

export const autocomplete = async (
  message: string,
  list: string[],
  args: Omit<inquirer.DistinctQuestion, 'type' | 'name'> = {},
): Promise<string> => {
  const fuse = new Fuse(list, {});

  return await prompt({
    ...args,
    type: 'autocomplete',
    message,
    pageSize: 20,
    source: (_: any, input: string) => {
      return !input ? list : fuse.search(input || '').map(({ item }) => item);
    },
  } as any);
};
