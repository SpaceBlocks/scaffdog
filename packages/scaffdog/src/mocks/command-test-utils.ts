import path from 'path';
import type {
  AnyCommandModule,
  CommandArgs,
  CommandModule,
  CommandOption,
  GlobalFlags,
  GlobalFlagsWith,
  InferredCommandModuleArgs,
  InferredCommandModuleFlags,
} from '../command';
import type { CommandContainer } from '../command-container';
import { createCommandContainer } from '../command-container';
import type { Library } from '../lib';
import { createLibraryMock } from './lib';
import { createLogger } from './logger';

export const cwd = path.resolve(__dirname, '../../');

export type CommandRunnerParams<T extends AnyCommandModule> = {
  args?: Partial<CommandArgs<InferredCommandModuleArgs<T>>>;
  flags?: Partial<GlobalFlagsWith<CommandArgs<InferredCommandModuleFlags<T>>>>;
  lib?: Library;
  container?: CommandContainer;
};

export type CommandRunner<T extends AnyCommandModule> = (
  params: CommandRunnerParams<T>,
) => Promise<{
  code: number;
  stdout: string;
  stderr: string;
}>;

export const createCommandRunner =
  <A extends CommandOption, F extends CommandOption>(
    cmd: CommandModule<A, F>,
    defaults: {
      args: CommandArgs<A>;
      flags: CommandArgs<F>;
    },
  ): CommandRunner<CommandModule<A, F>> =>
  async ({ args, flags, lib, container }) => {
    const { logger, getStdout, getStderr } = createLogger();

    const globalFlags: GlobalFlags = {
      project: '.scaffdog',
      help: false,
      version: false,
      verbose: true,
    };

    const code = await cmd.run({
      cwd,
      pkg: {
        name: 'scaffdog',
        version: '1.0.0',
        description: 'scaffdog is Markdown driven scaffolding tool.',
      },
      size: {
        rows: 120,
        columns: 60,
      },
      logger,
      container: container ?? createCommandContainer([]),
      lib: lib ?? createLibraryMock({}),
      args: {
        ...defaults.args,
        ...(args ?? {}),
      },
      flags: {
        ...defaults.flags,
        ...globalFlags,
        ...(flags ?? {}),
      } as never,
    });

    return {
      code,
      stdout: getStdout(),
      stderr: getStderr(),
    };
  };
