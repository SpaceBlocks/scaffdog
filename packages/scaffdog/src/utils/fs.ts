import fs from 'fs';
import { promisify } from 'util';

export const writeFile = promisify(fs.writeFile);
export const readFile = promisify(fs.readFile);
export const mkdir = promisify(fs.mkdir);

const fstat = (filepath: string) => {
  try {
    return fs.statSync(filepath);
  } catch (e) {
    return null;
  }
};

export const fileExists = (filepath: string): boolean =>
  fstat(filepath)?.isFile() ?? false;

export const directoryExists = (filepath: string): boolean =>
  fstat(filepath)?.isDirectory() ?? false;
