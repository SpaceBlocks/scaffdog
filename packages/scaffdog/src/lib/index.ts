import type { ConsolaInstance } from 'consola';
import { createConfigLibrary } from './config.js';
import { createDocumentLibrary } from './document.js';
import { createErrorLibrary } from './error.js';
import { createFsLibrary } from './fs.js';
import { createPromptLibrary } from './prompt.js';
import { createQuestionLibrary } from './question.js';

export const createLibrary = (logger: ConsolaInstance) => {
  const fs = createFsLibrary();
  const error = createErrorLibrary(logger);
  const prompt = createPromptLibrary();
  const config = createConfigLibrary(logger, error);
  const question = createQuestionLibrary(prompt);
  const document = createDocumentLibrary(fs, question);

  return {
    fs,
    error,
    prompt,
    config,
    question,
    document,
  };
};

export type Library = ReturnType<typeof createLibrary>;
