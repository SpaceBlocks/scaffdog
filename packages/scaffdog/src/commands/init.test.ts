import path from 'path';
import type { Mock } from 'vitest';
import { describe, expect, test, vi } from 'vitest';
import { createFsLibraryMock } from '../lib/fs.mock';
import { createPromptLibraryMock } from '../lib/prompt.mock';
import { createCommandRunner, cwd } from '../mocks/command-test-utils.js';
import { createLibraryMock } from '../mocks/lib.js';
import cmd from './init.js';

const run = createCommandRunner(cmd, {
  args: {},
  flags: {},
});

describe('success', () => {
  test('fresh', async () => {
    const lib = createLibraryMock({
      fs: createFsLibraryMock({
        directoryExists: vi.fn().mockReturnValue(false),
        mkdir: vi.fn().mockReturnValueOnce(Promise.resolve()),
        writeFile: vi.fn().mockReturnValue(Promise.resolve()),
      }),
      prompt: createPromptLibraryMock({
        prompt: vi.fn().mockResolvedValueOnce('pretty-dog'),
      }),
    });

    const { code, stdout, stderr } = await run({
      lib,
    });

    expect(code).toBe(0);
    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();

    expect(lib.fs.directoryExists).toBeCalledWith(
      path.resolve(cwd, '.scaffdog'),
    );

    expect(lib.fs.mkdir).toBeCalledWith(path.resolve(cwd, '.scaffdog'), {
      recursive: true,
    });

    const writeFile = lib.fs.writeFile as Mock;
    expect(writeFile.mock.calls.length).toBe(2);
    expect(path.relative(cwd, writeFile.mock.calls[0][0])).toBe(
      '.scaffdog/config.js',
    );
    expect(path.relative(cwd, writeFile.mock.calls[0][1])).toMatchSnapshot();
    expect(path.relative(cwd, writeFile.mock.calls[1][0])).toBe(
      '.scaffdog/pretty-dog.md',
    );
    expect(path.relative(cwd, writeFile.mock.calls[1][1])).toMatchSnapshot();
  });

  test('.scaffdog already exist', async () => {
    const lib = createLibraryMock({
      fs: createFsLibraryMock({
        directoryExists: vi.fn().mockReturnValue(true),
        mkdir: vi.fn().mockReturnValueOnce(Promise.resolve()),
        writeFile: vi.fn().mockReturnValue(Promise.resolve()),
      }),
      prompt: createPromptLibraryMock({
        prompt: vi.fn().mockResolvedValueOnce('pretty-dog'),
        confirm: vi.fn().mockResolvedValueOnce(true),
      }),
    });

    const { code, stdout, stderr } = await run({
      lib,
    });

    expect(code).toBe(0);
    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();
  });
});

describe('failure', () => {
  test('cancel', async () => {
    const lib = createLibraryMock({
      fs: createFsLibraryMock({
        directoryExists: vi.fn().mockReturnValue(true),
      }),
      prompt: createPromptLibraryMock({
        confirm: vi.fn().mockResolvedValueOnce(false),
      }),
    });

    const { code, stdout, stderr } = await run({
      lib,
    });

    expect(code).toBe(1);
    expect(stderr).toMatchSnapshot();
    expect(stdout).toBe('');
  });
});
