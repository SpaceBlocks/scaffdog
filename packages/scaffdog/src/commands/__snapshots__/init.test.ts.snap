// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`failure > cancel 1`] = `
"
 WARN  Setup canceled!

"
`;

exports[`success > .scaffdog already exist 1`] = `
"
Setup of scaffdog 🐶 is complete!

  ✔ .scaffdog/config.js
  ✔ .scaffdog/pretty-dog.md

Now you can do scaffold by running \`$ scaffdog generate\`.

Please refer to the following documents and customize it.
https://scaff.dog/docs/templates
"
`;

exports[`success > fresh 1`] = `
"
Setup of scaffdog 🐶 is complete!

  ✔ .scaffdog/config.js
  ✔ .scaffdog/pretty-dog.md

Now you can do scaffold by running \`$ scaffdog generate\`.

Please refer to the following documents and customize it.
https://scaff.dog/docs/templates
"
`;

exports[`success > fresh 2`] = `
"export default {
  files: ['*'],
};"
`;

exports[`success > fresh 3`] = `
"---
name: 'pretty-dog'
root: '.'
output: '**/*'
ignore: []
questions:
  value: 'Please enter any text.'
---

# \`{{ inputs.value }}.md\`

\`\`\`markdown
Let's make a document!
See scaffdog documentation for details.
https:/scaff.dog/docs/templates
\`\`\`"
`;
